app/src/main/java/org/wikipedia/language/addlanguages/AddLanguagesViewModel.kt

```diff
@@ -24,58 +25,65 @@ class AddLanguagesViewModel : ViewModel() {
     private val _siteInfoList = MutableStateFlow<List<SiteMatrix.SiteInfo>>(emptyList())
 
     // UI state exposed to Compose
-    private val _uiState = MutableStateFlow(LanguageListUiState())
+    private val _uiState = MutableStateFlow<LanguageListUiState>(LanguageListUiState.Loading)
     val uiState: StateFlow<LanguageListUiState> = _uiState.asStateFlow()
 
     private val handler = CoroutineExceptionHandler { _, throwable ->
         L.e(throwable)
-        _uiState.update {
-            it.copy(
-                error = throwable,
-                isSiteInfoLoaded = false
-            )
-        }
+        _uiState.value = LanguageListUiState.Error(throwable)
     }
 
+    private var fetchJob: Job? = null
+
     init {
       fetchAllData()
     }
 
     fun fetchAllData() {
-        // immediately populate with locally available data
-        populateInitialLanguageList()
-
-        // fetch site matrix
-        fetchSiteMatrix()
-    }
+        fetchJob?.cancel()
 
-    private fun populateInitialLanguageList() {
-        updateSearchTerm("", false)
-    }
+        // Set to loading state
+        _uiState.value = LanguageListUiState.Loading
 
-    private fun fetchSiteMatrix() {
-        viewModelScope.launch(handler) {
+        // fetch site matrix
+        fetchJob = viewModelScope.launch(handler) {
+            _uiState.value = LanguageListUiState.Loading
             val siteMatrix = ServiceFactory.get(WikipediaApp.instance.wikiSite).getSiteMatrix()
             val sites = SiteMatrix.getSites(siteMatrix)
             _siteInfoList.value = sites
 
-            // site is loaded so marking it
-            _uiState.update { it.copy(isSiteInfoLoaded = true) }
-
             // update the list
-            updateSearchTerm(_uiState.value.searchTerm, siteInfoAvailable = true)
+            if (isActive) {
```

Comment:
It would be great if you could add more comments for the usage of `isActive` here